pkg/services/hcloud/server/server.go

```diff
@@ -190,6 +217,198 @@ func (s *Service) handleBootStateUnset(ctx context.Context) (reconcile.Result, e
+		}
+
+		err = action.Error()
+		if err != nil {
+			err = fmt.Errorf("action %+v failed: %w", action, err)
+			log.FromContext(ctx).Error(err, "")
+			s.scope.SetError(err.Error(), capierrors.CreateMachineError)
+			return reconcile.Result{}, nil
+		}
+
+		log.FromContext(ctx).Info("Action RescueEnabled is finished",
+			"actionDuration", action.Finished.Sub(action.Started),
+			"finishedSince", time.Since(action.Finished))
+
+		hm.Status.ActionIDEnableRescueSystem = actionDone
+	}
+
+	action, err := s.scope.HCloudClient.Reboot(ctx, server)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("reboot failed: %w", err)
+	}
+
+	hm.Status.ActionIDRebootToRescue = action.ID
+
+	hm.SetBootState(infrav1.HCloudBootStateWaitForRescueRunningThenInstallImage)
+	return reconcile.Result{RequeueAfter: requeueAfterRebootToRescue}, nil
+}
+
+func (s *Service) handleBootStateWaitForRescueRunningThenInstallImage(ctx context.Context, _ *hcloud.Server) (reconcile.Result, error) {
+	hm := s.scope.HCloudMachine
+
+	if hm.Status.ActionIDRebootToRescue == 0 {
+		msg := "handleBootStateWaitForRescueRunningThenInstallImage ActionIdRebootToRescue not set? Can not continue. Provisioning Failed"
+		log.FromContext(ctx).Error(nil, msg)
+		s.scope.SetError(msg, capierrors.CreateMachineError)
+		return reconcile.Result{}, nil
+	}
+	if hm.Status.ActionIDRebootToRescue != actionDone {
+		action, err := s.scope.HCloudClient.GetAction(ctx, hm.Status.ActionIDRebootToRescue)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("GetAction failed: %w", err)
+		}
+
+		if action.Finished.IsZero() {
+			log.FromContext(ctx).Info("Waiting until Action RebootToRescue is finished", "action", action)
+			return reconcile.Result{RequeueAfter: requeueIntervalWaitForRescueRunning}, nil
+		}
+
+		err = action.Error()
+		if err != nil {
+			err = fmt.Errorf("action %+v failed: %w", action, err)
+			log.FromContext(ctx).Error(err, "")
+			s.scope.SetError(err.Error(), capierrors.CreateMachineError)
+			return reconcile.Result{}, nil
+		}
+
+		log.FromContext(ctx).Info("Action RebootToRescue is finished",
+			"actionDuration", action.Finished.Sub(action.Started),
+			"finishedSince", time.Since(action.Finished))
+
+		hm.Status.ActionIDRebootToRescue = actionDone
+	}
+
+	robotSecretName := s.scope.HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef.Name
+	if robotSecretName == "" {
+		return reconcile.Result{}, fmt.Errorf("HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef.Name is empty. Can not set hcloud machine into rescue-mode without ssh private key")
+	}
+
+	robotSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      robotSecretName,
+			Namespace: s.scope.Namespace(),
+		},
+	}
+
+	err := s.scope.Client.Get(ctx, client.ObjectKeyFromObject(robotSecret), robotSecret)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get secret %q: %w", robotSecretName, err)
+	}
+
+	if len(s.scope.HCloudMachine.Status.Addresses) == 0 {
```

Comment:
what should happen in this case? Is this temporary? Then we can requeue. Is this permanent? Then we have to set a condition